### PR TITLE
Feign & Docker & Kafka init

### PR DESCRIPTION
**Pull request by YevheniiChornyi**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -33,10 +33,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
     implementation 'org.springframework.boot:spring-boot-starter-jooq'
     implementation 'org.springframework.boot:spring-boot-starter-quartz'
-    implementation 'org.springframework.boot:spring-boot-starter-security'
-    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-web'
-    implementation 'org.thymeleaf.extras:thymeleaf-extras-springsecurity5'
+    implementation 'org.springframework.cloud:spring-cloud-starter-openfeign:3.1.0'
```
Comment (mmarusich) [inline build.gradle]: don't specify version, use spring dependency management

Comment (YevheniiChornyi) [reply build.gradle]: LostfilmDemo:test: Could not find org.springframework.cloud:spring-cloud-starter-openfeign:.
Required by:
    project :

File: src/main/java/yevhenii/lostfilmdemo/controllers/ImdbHolderClient.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import yevhenii.lostfilmdemo.entity.ImdbEpisodesHolder;
+import yevhenii.lostfilmdemo.entity.ImdbSeriesHolder;
+
+@FeignClient(url = "https://imdb-api.com/en/API/", name = "imdb")
+public interface ImdbHolderClient {
+
+    @GetMapping("/SearchSeries/k_vm88jkhr/{Id}")
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/controllers/ImdbHolderClient.java]: where this goes? check api docs http://www.omdbapi.com/

File: src/main/java/yevhenii/lostfilmdemo/entity/ImdbEpisodesHolder.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ImdbEpisodesHolder {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/entity/ImdbEpisodesHolder.java]: rename it to ImdbEpisodeDTO

File: src/main/java/yevhenii/lostfilmdemo/entity/ImdbSeriesHolder.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@NoArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ImdbSeriesHolder {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/entity/ImdbSeriesHolder.java]: same

File: src/main/java/yevhenii/lostfilmdemo/kafka/AlertReader.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class AlertReader {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/kafka/AlertReader.java]: what alert means?

File: src/main/java/yevhenii/lostfilmdemo/kafka/AlertSender.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class AlertSender {
+
+    private final KafkaTemplate<String, String> kafkaTemplate;
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/kafka/AlertSender.java]: send DTO, not a string

File: src/main/java/yevhenii/lostfilmdemo/rss/RssRunner.java [hunk 7]
```diff
@@ -2,19 +2,33 @@
...
 import org.springframework.boot.ApplicationArguments;
 import org.springframework.boot.ApplicationRunner;
 import org.springframework.stereotype.Component;
+import yevhenii.lostfilmdemo.controllers.ImdbHolderClient;
+import yevhenii.lostfilmdemo.kafka.AlertSender;
 
 @Slf4j
 @Component
 @RequiredArgsConstructor
 public class RssRunner implements ApplicationRunner {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/rss/RssRunner.java]: I think it's time to remove this and use quartz

File: build.gradle (hunk 1 above)
Comment (mmarusich) [reply build.gradle]: you need to add spring cloud dependency management config, you can generate new project using spring.initializer with feign and copy config from there

File: src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import yevhenii.lostfilmdemo.entity.ImdbEpisodesDTO;
+
+@FeignClient(url = "https://www.omdbapi.com", name = "omdb")
+public interface OmdbHolderClient {
+
+    @GetMapping("/?apikey={apiKey}&t={title}&Season={season}&Episode={episode}")
+    ImdbEpisodesDTO getEpisode(@PathVariable String apiKey, @PathVariable String title, @PathVariable int season, @PathVariable int episode);
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java]: why PathVariable? you using request params

File: src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package yevhenii.lostfilmdemo.controllers;
+
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import yevhenii.lostfilmdemo.entity.ImdbEpisodesDTO;
+
+@FeignClient(url = "https://www.omdbapi.com", name = "omdb")
+public interface OmdbHolderClient {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java]: just ImdbClient, why you adding `holder` into the name? what it holds?

File: src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+@RequiredArgsConstructor
+public class TVSeriesReader {
+    @Value("${omdb.api.key}")
+    private final String apiKey;
+    private final OmdbHolderClient omdbHolderClient;
+    private final TVSeriesRepository repository;
+    private final TVRecordConvertor convertor;
+
+    @KafkaListener(topics = {"${lostfilm.topic}"})
+    public void receive(ConsumerRecord<String, TVSeries> consumerRecord) {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java]: you can skip ConsumerRecord if you don't need key value, just change it to `receive(TVSeries record)`

File: src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        log.info("received: {}", consumerRecord.toString());
+        TVSeries tvSeries = createOmdbPart((consumerRecord.value()));
+        repository.update(convertor.createRecord(tvSeries));
+    }
+
+    private TVSeries createOmdbPart(TVSeries tvSeries) {
+
+        ImdbEpisodesDTO imdbSeries = omdbHolderClient.getEpisode(apiKey, tvSeries.getName(), tvSeries.getSeason(), tvSeries.getEpisode());
+        log.info(imdbSeries.toString());
+        if (imdbSeries.getImdbID() != null) {
```
Comment (mmarusich) [inline src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java:34]: how you can receive response from ImDB without id?

File: src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java (hunk 8 above)
Comment (mmarusich) [reply src/main/java/yevhenii/lostfilmdemo/controllers/OmdbHolderClient.java]: I also don't see in API @PathVariable int season, @PathVariable int episode supported

File: src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java (hunk 11 above)
Comment (YevheniiChornyi) [reply src/main/java/yevhenii/lostfilmdemo/kafka/TVSeriesReader.java:34]: Season+Episode search parameters added: http://www.omdbapi.com/?t=Game of Thrones&Season=1&Episode=1&apikey=3bb21e88

**Review by mmarusich — APPROVED** [review]

**Pull request merged**